framework/core/vulkan_resource.h

```diff
@@ -18,149 +19,267 @@
 	{
 		return handle;
 	}
-
-	inline const uint64_t get_handle_u64() const
+	else
 	{
-		// See https://github.com/KhronosGroup/Vulkan-Docs/issues/368 .
-		// Dispatchable and non-dispatchable handle types are *not* necessarily binary-compatible!
-		// Non-dispatchable handles _might_ be only 32-bit long. This is because, on 32-bit machines, they might be a typedef to a 32-bit pointer.
-		using UintHandle = typename std::conditional<sizeof(THandle) == sizeof(uint32_t), uint32_t, uint64_t>::type;
-
-		return static_cast<uint64_t>(reinterpret_cast<UintHandle>(handle));
+		return *reinterpret_cast<typename ResourceType::NativeType *>(&handle);
 	}
+}
 
-	inline const std::string &get_debug_name() const
+template <vkb::BindingType bindingType, typename Handle>
+inline const Handle &VulkanResource<bindingType, Handle>::get_handle() const
+{
+	if constexpr (bindingType == vkb::BindingType::Cpp)
 	{
-		return debug_name;
+		return handle;
+	}
+	else
+	{
+		return *reinterpret_cast<typename ResourceType::NativeType const *>(&handle);
 	}
+}
+
+template <vkb::BindingType bindingType, typename Handle>
+inline uint64_t VulkanResource<bindingType, Handle>::get_handle_u64() const
+{
+	// See https://github.com/KhronosGroup/Vulkan-Docs/issues/368 .
+	// Dispatchable and non-dispatchable handle types are *not* necessarily binary-compatible!
+	// Non-dispatchable handles _might_ be only 32-bit long. This is because, on 32-bit machines, they might be a typedef to a 32-bit pointer.
+	using UintHandle = typename std::conditional<sizeof(ResourceType) == sizeof(uint32_t), uint32_t, uint64_t>::type;
+
+	return static_cast<uint64_t>(*reinterpret_cast<UintHandle const *>(&handle));
+}
 
-	inline void set_debug_name(const std::string &name)
+template <vkb::BindingType bindingType, typename Handle>
+inline typename VulkanResource<bindingType, Handle>::ObjectType VulkanResource<bindingType, Handle>::get_object_type() const
+{
+	if constexpr (bindingType == vkb::BindingType::Cpp)
+	{
+		return ResourceType::objectType;
+	}
+	else
 	{
-		debug_name = name;
-		detail::set_debug_name(device, get_object_type(), get_handle_u64(), debug_name.c_str());
+		return static_cast<VkObjectType>(ResourceType::objectType);
 	}
+}
 
-  protected:
-	THandle     handle;
-	Device     *device;
-	std::string debug_name;
-};
+template <vkb::BindingType bindingType, typename Handle>
+inline bool VulkanResource<bindingType, Handle>::has_device() const
+{
+	return device != nullptr;
+}
 
+template <vkb::BindingType bindingType, typename Handle>
+inline bool VulkanResource<bindingType, Handle>::has_handle() const
+{
+	return handle != nullptr;
+}
+
+template <vkb::BindingType bindingType, typename Handle>
+inline void VulkanResource<bindingType, Handle>::set_debug_name(const std::string &name)
+{
+	debug_name = name;
+
+	if (device && !debug_name.empty())
+	{
+		get_device().get_debug_utils().set_debug_name(get_device().get_handle(), get_object_type(), get_handle_u64(), debug_name.c_str());
```

Comment:
Mostly cosmetic, but this looks very awkward. So many `get_` functions in here. Makes it kinda hard to read. I somehow prefer being able to directly access things like device, instance, atc. without having to use a getter.

Reply:
Unfortunately, directly accessing members of a class unifying C- and C++-bindings would not work, as those members should behave differently from C-based vs. C++-based code.
From C-based code, `get_handle()` gets you a `VkHandle`, while from C++-based code `get_handle()` gets you a `vk::Handle`. Internally (privately!), the handle is always stored as a `vk::Handle`, but your C-based code does not need to know that.

Reply:
Makes sense. Readability is still kinda odd though.

Reply:
One could introduce some `operator*()`, replacing the `get_handle()`, at least.
Would reduce the number of get-calls a little, but make it a bit less explicit.

Reply:
Yeah, maybe we should do that at a later point. Resolving this as I don't want to spam the PR ;)